flink-tests/src/test/java/org/apache/flink/test/recovery/BatchFineGrainedRecoveryITCase.java

```diff
@@ -101,84 +132,116 @@ public void teardown() throws Exception {
-			mapRestarts.clear();
-			expectedMapRestarts.clear();
-		}
+	private static class JoinedFailureStrategy implements FailureStrategy {
+		private static final long serialVersionUID = 1L;
 
-		private static int addNewMap() {
-			mapRestarts.add(new AtomicInteger(0));
-			expectedMapRestarts.add(new AtomicInteger(1));
-			return mapRestarts.size() - 1;
-		}
+		private final FailureStrategy[] failureStrategies;
 
-		private static void mapRestart(int index) {
-			mapRestarts.get(index).incrementAndGet();
+		private JoinedFailureStrategy(FailureStrategy ... failureStrategies) {
+			this.failureStrategies = failureStrategies;
 		}
 
-		private static void mapFailure(int index) {
-			expectedMapRestarts.get(index).incrementAndGet();
+		@Override
+		public boolean failOrNot(int trackingIndex) {
+			return Arrays
+				.stream(failureStrategies)
+				.anyMatch(failureStrategy -> failureStrategy.failOrNot(trackingIndex));
 		}
+	}
 
-		private static void verify() {
-			assertThat(collect(mapRestarts), is(collect(expectedMapRestarts)));
+	private static class RandomExceptionFailureStrategy extends AbstractRandomFailureStrategy {
+		private static final long serialVersionUID = 1L;
+
+		private RandomExceptionFailureStrategy(CoinToss coin) {
+			super(coin);
 		}
 
-		private static int[] collect(Collection<AtomicInteger> list) {
-			return list.stream().mapToInt(AtomicInteger::get).toArray();
+		@Override
+		void fail(int trackingIndex) {
+			StaticMapFailureTracker.mapFailure(trackingIndex);
+			throw new FlinkRuntimeException("BAGA-BOOM!!! The user function generated test failure.");
 		}
 	}
 
-	@FunctionalInterface
-	private interface FailureStrategy extends Serializable {
-		void failOrNot();
+	private static class RandomTaskExecutorFailureStrategy extends AbstractRandomFailureStrategy {
+		private static final long serialVersionUID = 1L;
+
+		private RandomTaskExecutorFailureStrategy(CoinToss coin) {
+			super(coin);
+		}
+
+		@Override
+		void fail(int trackingIndex) {
+			StaticMapFailureTracker.mapFailureWithBacktracking(trackingIndex);
+			try {
+				restartTaskManager();
+			} catch (InterruptedException e) {
+				Thread.currentThread().interrupt();
+			} catch (Exception e) {
+				ExceptionUtils.rethrow(e);
+			}
+		}
 	}
 
-	private static class RandomExceptionFailureStrategy implements FailureStrategy {
+	private abstract static class AbstractRandomFailureStrategy implements FailureStrategy {
 		private static final long serialVersionUID = 1L;
 
 		private final CoinToss coin;
 
-		private RandomExceptionFailureStrategy(int probFraction, int probBase) {
-			this.coin = new CoinToss(probFraction, probBase);
+		private AbstractRandomFailureStrategy(CoinToss coin) {
+			this.coin = coin;
 		}
 
 		@Override
-		public void failOrNot() {
-			if (coin.toss() && StaticFailureCounter.failOrNot()) {
-				throw new FlinkRuntimeException("BAGA-BOOM!!! The user function generated test failure.");
+		public boolean failOrNot(int trackingIndex) {
```

Comment:
What's the idea behind returning a boolean? The `RandomExceptionFailureStrategy` throws an error right away (), and the other kills the TM which should also result in a new instance being created (after all, there is only 1 TM).

In which cases would the return value actually be used?

EDIT: OK, so this is used in `JoinedFailureStrategy#failOrNot` to prevent multiple strategies from issuing failures. This is a fairly odd way of accomplishing this imo; I would prefer if we would separate the determination of whether we should fail from the actual failure.
Something like `boolean shouldFail()` and `void fail() throws Exception`, which you could then use in the `JoinedFailureStrategy` using `failureStrategies.anyMatch(FS::shouldFail).ifPresent(FS::fail)`.
You can still have a default `failOrNot` implementation that does both at once for easier use in the functions.

Reply:
I think we do not need 2 methods at the moment, one expresses the same thing.
I will add `throws Exception` to `failOrNot` and explain that it can fail either by throwing an exception or implicitly later and return `true` (TM failure).